src/gluonts/dataset/loader_v2.py

```diff
@@ -0,0 +1,189 @@
+        return list(stack(t, make_array_fn) for t in zip(*data))
+    return data
+
+
+def batchify(
+    data: List[dict], make_array_fn: Callable = np.asarray
+) -> DataBatch:
+    return {
+        key: stack(
+            data=[item[key] for item in data], make_array_fn=make_array_fn
+        )
+        for key in data[0].keys()
+    }
+
+
+class CyclicIterable(Iterable):
+    def __init__(self, base_iterable: Iterable) -> None:
+        self.base_iterable = base_iterable
+
+    def __iter__(self):
+        while True:
+            yield from self.base_iterable
+
+
+class PseudoShuffledIterable(Iterable):
+    def __init__(self, base_iterable: Iterable, buffer_length: int):
+        self.base_iterable = base_iterable
+        self.buffer_length = buffer_length
+        self.shuffle_buffer: list = []
+
+    def sample_from_buffer(self):
+        idx = random.randint(0, len(self.shuffle_buffer) - 1)
+        return self.shuffle_buffer.pop(idx)
+
+    def __iter__(self):
+        for x in self.base_iterable:
+            if len(self.shuffle_buffer) < self.buffer_length:
+                self.shuffle_buffer.append(x)
+            if len(self.shuffle_buffer) == self.buffer_length:
+                yield self.sample_from_buffer()
+        while len(self.shuffle_buffer):
+            yield self.sample_from_buffer()
+
+
+class MultiProcessIterator(Iterator):
+    def __init__(
+        self,
+        base_iterable: Iterable,
+        num_workers: int,
+        num_entries: int,
+        max_queue_size: Optional[int] = None,
+    ):
+        assert num_workers >= 1
+        assert num_entries >= 0
+        assert max_queue_size is None or max_queue_size >= num_workers
+
+        self.base_iterable = base_iterable
+        self.num_workers = num_workers
+        self.num_entries = num_entries
+        self.max_queue_size = (
+            max_queue_size if max_queue_size is not None else 5 * num_workers
+        )
+
+        self.data_queue: Queue = Queue(maxsize=self.max_queue_size)
+        self.done_event = Event()
+        self.processes = []
+
+        for wid in range(self.num_workers):
+            p = Process(
+                target=self.worker_fn,
+                args=(
+                    wid,
+                    self.num_workers,
+                    self.base_iterable,
+                    self.data_queue,
+                    self.done_event,
+                ),
+            )
+            p.start()
+            self.processes.append(p)
+
+        self.count = 0
+
+    @staticmethod
+    def worker_fn(
+        worker_id: int,
+        num_workers: int,
+        iterable: Iterable,
+        data_queue: Queue,
+        end_event,
+    ):
+        MPWorkerInfo.worker_process = True
+        MPWorkerInfo.worker_id = worker_id
+        MPWorkerInfo.num_workers = num_workers
+
+        for entry in iterable:
+            if end_event.is_set():
+                break
+            data_queue.put((worker_id, entry))
```

Comment:
I'm concerned that putting the entries directly into the queue is too slow, essentially this is the kind of thing we would like to prevent in multiprocessing since its a bottleneck that requires synchronisation between processes and is capped by slow access speeds of the RAM.
- as @lostella mentioned, this could be partially circumvented by having a unique Queue per process
- sill, pickling the time series data into the Queue is not a good idea in general imo
- a more elegant way to do it is pickling the data into shared memory and communicating the address through the queue: like done here: https://github.com/awslabs/gluon-ts/blob/3774b1df6d2b9fb75fbabc181fadb188c34fc6db/src/gluonts/dataset/parallelized_loader.py#L353

Reply:
This doesn't appear to be an issue, but I've only quickly benchmarked this: will do more thorough testing